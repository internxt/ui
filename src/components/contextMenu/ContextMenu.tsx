import Menu, { MenuItemsType } from '../menu/Menu';

const MENU_BUTTON_HEIGHT = 40;

export interface ContextMenuProps<T> {
  item: T;
  menuItemsRef: React.MutableRefObject<HTMLDivElement | null>;
  menu?: MenuItemsType<T>;
  openedFromRightClick: boolean;
  posX: number;
  posY: number;
  isContextMenuCutOff: boolean;
  genericEnterKey: () => void;
  handleMenuClose: () => void;
}

/**
 * Properties for `ContextMenuProps<T>`
 *
 * @template T - Generic type representing the item to which the context menu applies.
 *
 * @property {T} item
 * - The current item associated with the context menu.
 * This object is passed to each menu option's `action` and `disabled` functions.
 *
 * @property {React.MutableRefObject<HTMLDivElement | null>} menuItemsRef
 * - A mutable ref to the `div` containing the context menu. Used for handling the menu's positioning and visibility.
 *
 * @property {MenuType<T>} [menu]
 * - An array of menu options, where each option includes properties like `name`, `icon`, `action`, etc.
 * The `MenuType<T>` type allows some options to be separators (`separator: boolean`).
 *
 * @property {boolean} openedFromRightClick
 * - Indicates whether the context menu was opened via a right-click (`true`).
 * Determines whether the menu's position is set based on the click location or a predefined position.
 *
 * @property {number} posX
 * - X-coordinate for the menu's position, used if `openedFromRightClick` is `true`.
 *
 * @property {number} posY
 * - Y-coordinate for the menu's position, used if `openedFromRightClick` is `true`.
 *
 * @property {boolean} isContextMenuCutOff
 * - Specifies whether the menu should align to the bottom of the screen to prevent it
 * from being cut off on smaller screens. Switches menu positioning between top or bottom.
 *
 * @property {() => void} genericEnterKey
 * - A callback that executes if the Enter key is pressed without a selected menu option.
 * Can be used to define a default action.
 *
 * @property {() => void} handleMenuClose
 * - Function to close the context menu. Called after an action is executed or when pressing Enter in the menu.
 */

const ContextMenu = <T,>({
  item,
  menuItemsRef,
  menu,
  openedFromRightClick,
  posX,
  posY,
  isContextMenuCutOff,
  genericEnterKey,
  handleMenuClose,
}: ContextMenuProps<T>): JSX.Element => {
  return (
    <div
      className="z-20 mt-0 flex flex-col rounded-lg bg-surface py-1.5 shadow-subtle-hard outline-none dark:bg-gray-5"
      style={
        openedFromRightClick
          ? { position: 'absolute', left: posX, top: posY, zIndex: 99 }
          : {
              position: 'absolute',
              right: 20,
              [isContextMenuCutOff ? 'bottom' : 'top']: MENU_BUTTON_HEIGHT,
              zIndex: 9999,
            }
      }
      ref={menuItemsRef}
    >
      <Menu item={item} genericEnterKey={genericEnterKey} handleMenuClose={handleMenuClose} menu={menu} />
    </div>
  );
};

export default ContextMenu;
