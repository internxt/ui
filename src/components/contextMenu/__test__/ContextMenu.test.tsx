import { render, fireEvent, screen } from '@testing-library/react';
import ContextMenu, { ContextMenuProps, MenuItemType } from '../ContextMenu';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import React from 'react';

interface TestItem {
  name: string;
}

describe('ContextMenu Component', () => {
  const props: ContextMenuProps<TestItem> = {
    item: { name: 'Test Item' },
    menuItemsRef: { current: null },
    menu: [
      { name: 'Option 1', action: vi.fn() },
      { name: 'Option 2', action: vi.fn(), disabled: () => false },
      { separator: true },
      { name: 'Option 3', action: vi.fn() },
    ],
    openedFromRightClick: true,
    posX: 100,
    posY: 100,
    isContextMenuCutOff: false,
    genericEnterKey: vi.fn(),
    handleMenuClose: vi.fn(),
  };

  const renderContextMenu = () => {
    return render(<ContextMenu {...props} />);
  };

  beforeEach(() => {});

  it('should match snapshot', () => {
    const contextMenu = renderContextMenu();
    expect(contextMenu).toMatchSnapshot();
  });

  it('renders menu items correctly', () => {
    renderContextMenu();
    expect(screen.getByText('Option 1')).toBeInTheDocument();
    expect(screen.getByText('Option 2')).toBeInTheDocument();
    expect(screen.getByText('Option 3')).toBeInTheDocument();
  });

  it('calls action and handleMenuClose on item click', () => {
    render(<ContextMenu {...props} />);
    const option1 = screen.getByText('Option 1');
    fireEvent.click(option1);
    const menuItem = props.menu?.[0];
    expect(menuItem && 'action' in menuItem).toBeTruthy();
    if (menuItem && 'action' in menuItem) expect(menuItem?.action).toHaveBeenCalledWith(props.item);
    expect(props.handleMenuClose).toHaveBeenCalled();
  });

  it('does not call action if the item is disabled', () => {
    const disabledMenu: MenuItemType<TestItem> = {
      name: 'Disabled Option',
      action: vi.fn(),
      disabled: () => true,
    };
    props.menu?.push(disabledMenu);
    render(<ContextMenu {...props} />);
    const disabledOption = screen.getByText('Disabled Option');
    fireEvent.click(disabledOption);
    expect(disabledMenu.action).not.toHaveBeenCalled();
  });

  it('calls genericEnterKey when Enter is pressed without selection', () => {
    render(<ContextMenu {...props} />);
    fireEvent.keyDown(window, { key: 'Enter' });
    expect(props.genericEnterKey).toHaveBeenCalled();
    expect(props.handleMenuClose).toHaveBeenCalled();
  });

  it('navigates through items with ArrowDown and ArrowUp keys', () => {
    render(<ContextMenu {...props} />);
    fireEvent.keyDown(window, { key: 'ArrowDown' });
    expect(screen.getByText('Option 1').parentElement?.parentElement).toHaveClass('dark:bg-gray-10');
    fireEvent.keyDown(window, { key: 'ArrowDown' });
    expect(screen.getByText('Option 2').parentElement?.parentElement).toHaveClass('dark:bg-gray-10');
    fireEvent.keyDown(window, { key: 'ArrowUp' });
    expect(screen.getByText('Option 1').parentElement?.parentElement).toHaveClass('dark:bg-gray-10');
  });

  it('calls action and handleMenuClose on Enter key with an item selected', () => {
    render(<ContextMenu {...props} />);
    fireEvent.keyDown(window, { key: 'ArrowDown' });
    fireEvent.keyDown(window, { key: 'Enter' });
    const menuItem = props.menu?.[0];
    expect(menuItem && 'action' in menuItem).toBeTruthy();
    if (menuItem && 'action' in menuItem) expect(menuItem?.action).toHaveBeenCalledWith(props.item);
    expect(props.handleMenuClose).toHaveBeenCalled();
  });
});
