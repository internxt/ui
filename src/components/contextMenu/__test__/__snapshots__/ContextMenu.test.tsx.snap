// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContextMenu Component > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="z-20 mt-0 flex flex-col rounded-lg bg-surface py-1.5 shadow-subtle-hard outline-none dark:bg-gray-5"
        style="position: absolute; left: 100px; top: 100px; z-index: 99;"
      >
        <div
          class="z-20 mt-0 flex flex-col rounded-lg bg-surface py-1.5 shadow-subtle-hard outline-none dark:bg-gray-5"
        >
          <div>
            <div
              class=""
            >
              <div
                class="flex cursor-pointer flex-row whitespace-nowrap px-4 py-1.5 text-base
                    
                    
                    
                    text-gray-80
                  "
              >
                <div
                  class="flex flex-row items-center space-x-2"
                >
                  <span>
                    Option 1
                  </span>
                </div>
                <span
                  class="ml-5 flex grow items-center justify-end text-sm text-gray-40"
                />
              </div>
            </div>
          </div>
          <div>
            <div
              class=""
            >
              <div
                class="flex cursor-pointer flex-row whitespace-nowrap px-4 py-1.5 text-base
                    
                    
                    
                    text-gray-80
                  "
              >
                <div
                  class="flex flex-row items-center space-x-2"
                >
                  <span>
                    Option 2
                  </span>
                </div>
                <span
                  class="ml-5 flex grow items-center justify-end text-sm text-gray-40"
                />
              </div>
            </div>
          </div>
          <div>
            <div
              class="my-0.5 flex w-full flex-row px-4"
            >
              <div
                class="h-px w-full bg-gray-10"
              />
            </div>
          </div>
          <div>
            <div
              class=""
            >
              <div
                class="flex cursor-pointer flex-row whitespace-nowrap px-4 py-1.5 text-base
                    
                    
                    
                    text-gray-80
                  "
              >
                <div
                  class="flex flex-row items-center space-x-2"
                >
                  <span>
                    Option 3
                  </span>
                </div>
                <span
                  class="ml-5 flex grow items-center justify-end text-sm text-gray-40"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="z-20 mt-0 flex flex-col rounded-lg bg-surface py-1.5 shadow-subtle-hard outline-none dark:bg-gray-5"
      style="position: absolute; left: 100px; top: 100px; z-index: 99;"
    >
      <div
        class="z-20 mt-0 flex flex-col rounded-lg bg-surface py-1.5 shadow-subtle-hard outline-none dark:bg-gray-5"
      >
        <div>
          <div
            class=""
          >
            <div
              class="flex cursor-pointer flex-row whitespace-nowrap px-4 py-1.5 text-base
                    
                    
                    
                    text-gray-80
                  "
            >
              <div
                class="flex flex-row items-center space-x-2"
              >
                <span>
                  Option 1
                </span>
              </div>
              <span
                class="ml-5 flex grow items-center justify-end text-sm text-gray-40"
              />
            </div>
          </div>
        </div>
        <div>
          <div
            class=""
          >
            <div
              class="flex cursor-pointer flex-row whitespace-nowrap px-4 py-1.5 text-base
                    
                    
                    
                    text-gray-80
                  "
            >
              <div
                class="flex flex-row items-center space-x-2"
              >
                <span>
                  Option 2
                </span>
              </div>
              <span
                class="ml-5 flex grow items-center justify-end text-sm text-gray-40"
              />
            </div>
          </div>
        </div>
        <div>
          <div
            class="my-0.5 flex w-full flex-row px-4"
          >
            <div
              class="h-px w-full bg-gray-10"
            />
          </div>
        </div>
        <div>
          <div
            class=""
          >
            <div
              class="flex cursor-pointer flex-row whitespace-nowrap px-4 py-1.5 text-base
                    
                    
                    
                    text-gray-80
                  "
            >
              <div
                class="flex flex-row items-center space-x-2"
              >
                <span>
                  Option 3
                </span>
              </div>
              <span
                class="ml-5 flex grow items-center justify-end text-sm text-gray-40"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
